src/mcts/search.cc

```diff
@@ -478,12 +482,12 @@ std::vector<EdgeAndNode> Search::GetBestChildrenNoTemperature(Node* parent,
     PopulateRootMoveLimit(&root_limit);
   }
   // Best child is selected using the following criteria:
-  // * Is terminal win, e.g., checkmate.
+  // * Prefer shorter terminal wins / avoid shorter terminal losses.
   // * Largest number of playouts.
   // * If two nodes have equal number:
   //   * If that number is 0, the one with larger prior wins.
   //   * If that number is larger than 0, the one with larger eval wins.
-  using El = std::tuple<bool, uint64_t, float, float, EdgeAndNode>;
+  using El = std::tuple<float, uint64_t, float, float, EdgeAndNode>;
```

Comment:
That tuple is becoming so large that it probably makes sense to replace it with a proper struct (with custom `operator<`).
It will probably make the code cleaner because there will be no need to invent values like terminal_rank.

Reply:
Actually it's likely that just a comparator of `EdgeAndNode` would be enough, without need to store any temporary values.

Reply:
Should we leave the `float terminal_rank` temporary value computation outside of `EdgeAndNode::operator<` as there is a little bit of rank computation for terminal nodes that might benefit from a temporary value instead of computing it on each `operator<` call.

Reply:
This function is only called once per 5 seconds, so it would be fine just with the comparator I guess.. There's not that many math there.

But if you think it makes sense to cache `terminal_rank` then maybe then use `struct`, something like this? :
```cpp
class EdgeRank {
 public:
  EdgeRank(EdgeAndNode* edge) : edge_(edge) {
    checkmate_distance_ = /* compute it */;
  }
  bool operator<(const EdgeRank& other) {
    if (checkmate_distance_ != other.checkmate_distance_) {
      return /* based on some logic */;
    }
    if (edge.GetN() != other.edge.GetN()) return edge.GetN() > other.edge.GetN();  // '>' because we'll want to sort it in descending order.
    // And so on..
  }
 private:
  EdgeAndNode edge_;
  int checkmate_distance_;  // 0 if no checkmate, negative if losing, positive if winning, +100 if tablebase.
};

Reply:
Or, I think I see what you're suggesting of not having a `terminal_rank` to begin with and have those details in the operator:
```c
bool EdgeAndNode::operator<(other) {
  if (IsTerminal() && other.IsTerminal()) {
    if (WL > 0 && other.WL > 0) return shorter winning mate;
…
```

Reply:
Yep.
But please not in `operator<` of `EdgeAndNode` but rather an external comparator. :)
Either
```cpp
struct EdgeRankComp{
   bool operator()(const EdgeAndNode& a, const EdgeAndNode& b) const {
    /* compare a and b */
}};
```
or a wrapper struct as I mentioned in the previous comment.

Or the comparator can be actually inline lambda in std::partial_sort:
```cpp
std::partial_sort(edges.begin(), middle, edges.end(), [](const auto& a, const auto& b) {
  // Compare!
});
```

Reply:
I believe https://github.com/LeelaChessZero/lc0/pull/1111/commits/7115097043a443852d6d3ee2561076f7702edd11 implements the explicit comparisons instead of relying on `float terminal_rank` with the appropriate cascading comparisons for N, WL, P.